addons/point_of_sale/static/src/app/generic_components/input/input.js

```diff
@@ -0,0 +1,73 @@
+/** @odoo-module */
+
+import { Component, useRef, useState } from "@odoo/owl";
+import { useAutofocus } from "@web/core/utils/hooks";
+import { debounce } from "@web/core/utils/timing";
+
+/**
+    This component is meant to provide a "batteries included" api for working
+    with inputs. It is well suited to work as a search bar or as a monetary input.
+    Optional props allow handling debouncing, toggling between mobile and desktop views,
+    autofocus, validation, and more. 
+    The only required prop is tModel, which is an array that represents the
+    state variable that this component should read and write to. The first
+    element of the array is the object returned by `useState` and the rest of 
+    the elements are strings representing the keys to access the value.
+    (they can also be numbers in the case of arrays).
+    ex: normally you would write
+    ```<input t-model="state.userValues[nThValue].searchString"/>```
+    with this component you can write
+    ```<Input tModel="[state, "userValues", nThValue, "searchString"]"/>```
+*/
+export class Input extends Component {
+    static template = "point_of_sale.input";
+    static props = {
+        tModel: { type: Array },
+        isSmall: { type: Boolean, optional: true },
+        debounceMillis: { type: Number, optional: true },
+        icon: {
+            type: Object,
+            optional: true,
+            shape: { type: String, value: String },
+        },
+        getRef: { type: Function, optional: true },
+        autofocus: { type: Boolean, optional: true },
+        autofocusMobile: { type: Boolean, optional: true },
+        iconOnLeftSide: { type: Boolean, optional: true },
+        isValid: { type: Function, optional: true },
+        placeholder: { type: String, optional: true },
+        class: { type: String, optional: true },
+        callback: { type: Function, optional: true },
+    };
+    static defaultProps = {
+        class: "",
+        isSmall: false,
+        debounceMillis: 0,
+        icon: {},
+        placeholder: "",
+        autofocus: false,
+        autofocusMobile: false,
+        iconOnLeftSide: true,
+        isValid: () => true,
+    };
+    setup() {
+        this.state = useState({
+            isOpen: false,
+        });
+        this.setValue = debounce(this.setValue, this.props.debounceMillis);
+        this.props.getRef?.(
+            (this.props.autofocus &&
+                useAutofocus({ refName: "input", mobile: this.props.autofocusMobile })) ||
+                useRef("input")
+        );
+    }
+    getValue(tModel = this.props.tModel) {
+        const [obj, ...keyPath] = tModel;
+        return keyPath.reduce((value, key) => value[key], obj);
+    }
+    setValue(newValue, tModel = this.props.tModel) {
+        this.props.callback?.(newValue);
+        const [obj, ...keyPath] = tModel;
+        keyPath.slice(0, -1).reduce((value, key) => value[key], obj)[keyPath.at(-1)] = newValue;
+    }
```

Comment:
Looks to be more readable if we do like this. Also, in my opinion we should always set the value even if the callback crashes.
```suggestion
    setValue(newValue) {
        try {
            this.props.callback?.(newValue);
        } finally {
            const tModel = this.props.tModel;
            const [obj, ...keyPath] = tModel;
            const container = this.getValue([obj, ...keyPath.slice(0, -1)]);
            container[keyPath.at(-1)] = newValue;
        }
    }
```